### PR TITLE
Develop docupdate

### PR DESCRIPTION
**Pull request by sebastiansam55**
Opening this as a draft to spur conversation on the best way to update/modernize the documentation. 

There are a number of things that need additional polish before final commit,
- [x] Change the edit on github to reflect autokey
- [ ] Wiki conversion if that is something that we are interested in
  - Benefit of having the wiki here would be that the changes would be easier to track/manage but it would remove the ability for small edits that probably don't need that much oversight.
- [ ] Multiversion
  - There are extensions that allow you to maintain different documentation versions, I have looked into it a bit but I'm not sure how much benefit it would be as I assume we'd like to keep most people on the latest version when possible.
  - I think that we should probably simplify this to having `main` and `develop` versions
  - This will work out to the "main" documentation being in the base `autokey.github.io/` and the develop being put in `autokey.github.io/dev/`
  - Might be a banner we could add to the top of the page to inform people they are not using the main doc
- [x] The main .gitignore includes html files
- [x] Swap image files to link to prexisting versions in the repo if possible
- [ ] Add documentation for the documentation
- [ ] Convert old epytext style documentation to sphinx syntax (long term, not really that valuable currently)
- [x] Update the readme in the doc/sphinx folder
- [ ] Add a Github workflow

I believe adding it to the "main" autokey repo instead of the `autokey.github.io` will be a benefit as it will be closer to where the development happens, and hopefully stay more updated because of this. It also appears to be relatively trivial to implement automatic updates to the documentation https://github.com/marketplace/actions/sphinx-build 

Simple things like adding new methods to API libraries and updating documentation would be able to be added to the doc site automatically.

Obviously open to any suggestions/comments/concerns.

I have based these changes on the develop branch as the refactor makes backwards compatibility more difficult to obtain.

You can see what the end result would be here https://samuelsebastian.dev/autokey/api.html

### CODE REVIEW

**Comment by kreezxil** [comment]
Sounds good to me.

**Comment by josephj11** [comment]
Lovely. Still hoping that visgrep will get added eventually.

**Comment by sebastiansam55** [comment]
@josephj11 that exists! It's under the high-level documentation page!

**Comment by josephj11** [comment]
Cool! I had no idea to look for it there.

**Review by sebastiansam55 — COMMENTED** [review]
@BlueDrink9 Not sure if it would be more appropriate to have these changes (only to code comments/documentation!) in this PR or in a separate PR, I have included them here for the time being.

I was trying to leave this comment on a specific commit but I guess it doesn't work like that, I was referring to this commit:
https://github.com/autokey/autokey/pull/538/commits/27d59440e0f307d0268599b17549fa92626d783c

Edit: and I'll squash most of these commits whenever we end up merging (hopefully at the same time as release of the new version!)

**Comment by BlueDrink9** [comment]
Initial thoughts (other than this being brilliant):

Multi-version docs are useful, as is having docs in the main repo rather than .io,  because it makes it easier for devs to document new features in the same PR that they submit them in. That's really helpful for keeping timely and accurate docs.

**Comment by BlueDrink9** [comment]
I think we should keep the wiki as-is. Perhaps integrate the more official documentation of installation and core functions into the repo, but leave the wiki for user tips etc.

**Comment by josephj11** [comment]
Put this where it works best for you to keep it up to date. I think some people didn't find the github.io stuff because there was nothing obvious pointing to it, so if it's moved into the regular repo, that will probably help people to find it.

I assume this won't affect the rest of the wiki. It's great being able to edit it without dealing with git directly, etc.

**Comment by sebastiansam55** [comment]
Hope I did the git stuff right or whatever. `tox -e docs` builds the documentation as expected now! Very cool.

**Comment by BlueDrink9** [comment]
Oh wow, cool! Very convenient. Once we sort out where this stuff is going to end up, I'll look into including updating docs as part of the new `build` CI workflow.

Regarding git stuff, best practise is to `git rebase upstream/develop` or something similar. Not sure how you have included these commits here. It doesn't look like they'll conflict too much, but it helps the history to rebase. (Rebasing basically replays your commits on top of the branch you specify, as if you had originally started  your work from there).

Just to update, what is the status of the new docs? What decisions/work is holding it up, how complete is it?

**Comment by sebastiansam55** [comment]
ack, tried to rebase but must have done it wrong

**Comment by BlueDrink9** [comment]
I'll rebase it for you, probably easier. For now, undo your rebase using `reset` and the `reflog`

**Comment by sebastiansam55** [comment]
I think I merged the upstream develop changes since I last worked on this, and regenerated the documentation for the latest commit.

@BlueDrink9 It's ready to be merged IMO, we just need to setup this repository to use github pages and the other documentation repo to be disabled.

This site is the github pages for the branch on my fork, this is what it will look like when published;
https://samuelsebastian.dev/autokey/index.html

**Comment by BlueDrink9** [comment]
@josephj11 we're probably going to want to make a list of places we need to update that link to existing documentation

**Comment by BlueDrink9** [comment]
And figure out what will be needed for a pages branch. 

Actually shoot, I think this may have been prematurely merged. I note that we have lost a couple of sections from contributers.RST! Could you please open a new PR to fix them?

**Comment by BlueDrink9** [comment]
Also we now have two folders: `doc` and `docs`. doc is the source, and `docs` are the generated docs. Perhaps we should set up CI to automatically regenerate the docs into a different repo?

I'd like a way to view documentation for multiple versions, such is common on readthedocs pages

**Comment by BlueDrink9** [comment]
Anyway, new docs are currently live at https://autokey.github.io/autokey/

**Comment by BlueDrink9** [comment]
@josephj11 you will need to go into the settings at https://github.com/autokey/autokey.github.io and archive it, and disable gh-pages

**Comment by BlueDrink9** [comment]
Ok yes merging this was a mistake. We seem to have ended up with a duplicate of the actions yml, probably because of a rename. Don't have time to fix it atm, but until we do, tests will be broken

File: lib/autokey/scripting/window.py [hunk 1]
```diff
@@ -71,7 +61,7 @@ def wait_for_focus(self, title, timeOut=5):
 
         return False
 
-    def wait_for_exist(self, title, timeOut=5, by_hex=False):
+    def wait_for_exist(self, title, timeOut=5):
```
Comment (BlueDrink9) [inline lib/autokey/scripting/window.py:64]: Why is code changed here if this is just a documentation update?

**Comment by BlueDrink9** [comment]
Reverted for now. Might have to re-create this PR sorry, but this time I will check it over properly! Eek. Sorry Sam.

**Comment by BlueDrink9** [comment]
Make sure to merge the latest `develop` into your branch when you recreate the PR

**Comment by sebastiansam55** [comment]
I thought that is what I did, I merged upstream develop into my develop-docupdate. I'll work on getting a new PR ready

**Comment by sebastiansam55** [comment]
So it seems like there is one commit responsible for the window.py changes, will reverting that plus deleting the extra python-app.yml resolve the issues here?

https://github.com/autokey/autokey/pull/538/commits/3b2eae16f76e8f8753f171aef246e9f67c540dd9

Sorry for not catching that, it removes code that I wrote for `window.py` lol. For what it's worth I went through the rest of the .py files and the rest of the changes are limited to the comment updates, minus one or two files where it added a newline at the end of the file.

**Comment by BlueDrink9** [comment]
Run a diff between the old `yml` and the new one. I suspect your changes to the `yml` might be lost in the new one? You may have to manually transfer them to the new one.

Regarding the merge: ok, sweet. Just needed to look more closely at the result of it. Perhaps a rebase on top of develop would have been more appropriate, but at this stage you're starting to look at git-fu with the reflog and picking up commits. I think I could do it from your machine, but idk what your git confidence is like

**Comment by BlueDrink9** [comment]
A revert of that commit would work, but isn't as clean as an interactive rebase that `drop`s that commit. `git rebase -i 3b2eae1`

**Comment by josephj11** [comment]
> @josephj11 you will need to go into the settings at https://github.com/autokey/autokey.github.io and archive it, and disable gh-pages

I'm not sure what this is about and I don't want to break anything.

Once this is resolved, I assume we can do some sort of web search - like using the search engine `site:` predicate to find old broken links. But II'm not sure on that either.

**Comment by Elliria** [comment]
Okay, I reread all the messages from @userdocs and @BlueDrink9 and looked at the referenced PRs and reverts, etc., and after puzzling and puzzling over it, I think I understand most of it:

The AutoKey project has [three repositories](https://github.com/autokey):

1. [autokey-python2](https://github.com/autokey/autokey-python2) - the old AutoKey Python 2 repository
2. [autokey.github.io](https://github.com/autokey/autokey.github.io) - the old documentation for AutoKey 0.95.10 and earlier
3. [autokey](https://github.com/autokey/autokey) - our current AutoKey repository
****
The files in [/doc/scripting](https://github.com/autokey/autokey/tree/master/doc/scripting) in our current repository are more recent versions of the files in [the old documentation repository](https://github.com/autokey/autokey.github.io).

What I don't understand is the purpose for the more-recent versions of the old document files in our current repository. If you [visit the old documentation](https://autokey.github.io/index.html), you can see that it was generated in 2012, which is way before any of the dates of any of those files, so they're not actually being used.
****
I think @BlueDrink was suggesting that in order to retire the old documentation (which we don't want to do yet since that's still in use by folks who are using older releases), you would archive [the files in the old documentation repository](https://github.com/autokey/autokey.github.io) and then [disable GitHub pages](https://docs.github.com/en/pages/getting-started-with-github-pages/unpublishing-a-github-pages-site) so they will no longer be generated. 

It's possible that, depending on the purpose for the files in [/doc/scripting](https://github.com/autokey/autokey/tree/master/doc/scripting) in our current repository, those should also be archived along with the original files and then removed from the current AutoKey repository once the complete retirement of the old documentation is official. Hopefully someone who's familiar with their purpose will chime in here (or you and I will have to simply stare at those docs, @josephj11, until the reason for their existence becomes evident).
****
The current official documentation is [here](https://autokey.github.io/autokey/), but as @userdocs pointed out, it can't be updated because `/docs` is missing from our current repository. That also explains why we get [a 404 page like this one](https://github.com/autokey/autokey/blob/docs/index.rst) when we click the **Edit on GitHub** link in the upper right corner of any of the pages in the current official documentation. It also explains why you're seeing some sort of error when you're merging PRs.

Can we just create `/docs` to make it so the current official documentation will be automatically built from now on and possibly also will be editable from the **Edit on GitHub** links on its pages? Or is there more involved in getting this taken care of?
****
By the way, @josephj11, if you'd like to put a link to the current official documentation in the right panel on AutoKey's main GitHub page, go to [https://github.com/autokey/autokey](https://github.com/autokey/autokey), click the little gear next to **About** in the right panel, paste the [https://autokey.github.io/autokey](https://autokey.github.io/autokey) URL into the **Website** box, and save the change.

**Comment by BlueDrink9** [comment]
Updated stuff: probably manually updated, but never got moved.

Creating /docs won't automatically generate new docs. Other way around: generating new docs will create /docs

**Comment by Elliria** [comment]
Oh! How do we do that?

**Comment by josephj11** [comment]
When you get this sufficiently unravelled, I'll be happy to stare at any questionable files so we can clean this up.

Did the "About" change. Good idea.

**Comment by Elliria** [comment]
You and I are probably going to be doing a lot of staring as time goes by. I'm slowly coming to understand this stuff, but it's taken quite a bit of staring already in combination with occasional lengthy discussions with my little Pikachu "rubber ducky" and I have no doubt there will be much more to come.

The "About" change looks good. I did it to my dummy repository as well after figuring out GitHub Pages today.

**Pull request merged**